src/Http/Routing/src/Builder/RouteHandlerFilterExtensions.cs

```diff
@@ -32,11 +32,22 @@ public static RouteHandlerBuilder AddFilter(this RouteHandlerBuilder builder, IR
     /// <returns>A <see cref="RouteHandlerBuilder"/> that can be used to further customize the route handler.</returns>
     public static RouteHandlerBuilder AddFilter<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TFilterType>(this RouteHandlerBuilder builder) where TFilterType : IRouteHandlerFilter
     {
-        var filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), Type.EmptyTypes);
+        // We call `CreateFactory` twice here since the `CreateFactory` API does not support optional arguments.
+        // See https://github.com/dotnet/runtime/issues/67309 for more info.
+        ObjectFactory filterFactory;
+        try
+        {
+            filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), new[] { typeof(RouteHandlerContext) });
+        }
+        catch (InvalidOperationException)
+        {
+            filterFactory = ActivatorUtilities.CreateFactory(typeof(TFilterType), Type.EmptyTypes);
+        }
+
         builder.RouteHandlerFilterFactories.Add((routeHandlerContext, next) => (context) =>
         {
-            var filter = (IRouteHandlerFilter)filterFactory.Invoke(context.HttpContext.RequestServices, Array.Empty<object>());
-            return filter.InvokeAsync(context, next, routeHandlerContext);
+            var filter = (IRouteHandlerFilter)filterFactory.Invoke(context.HttpContext.RequestServices, new[] { routeHandlerContext });
```

Comment:
Is this is called per request? If so the `new[] { routeHandlerContext }` should be moved into the out lambda.